libs/filter_example/src/example_filter.rs

```diff
@@ -52,213 +95,364 @@ impl Filter {
         for key in string_data.keys() {
             hash.insert(key.clone(), State::new_with_str(string_data[key].clone()));
         }
-        Box::into_raw(Box::new(Filter { filter_state: hash }))
+        Box::into_raw(Box::new(Filter {
+            whoami: None,
+            target_graph: None,
+            filter_state: hash,
+            envoy_shared_data: HashMap::new(),
+            collected_properties: vec!["service_name".to_string(), "height".to_string()],
+        }))
     }
 
-    #[no_mangle]
-    pub fn execute(&mut self, x: &Rpc) -> Vec<Rpc> {
-        let rpc_to_return = Rpc {
-            data: x.data.clone(),
-            uid: x.uid,
-            headers: x.headers.clone(),
-        };
-        let mut to_return = vec![rpc_to_return];
-        let mod_rpc = &mut to_return[0];
-        // 0. Who am I?
+    pub fn init_filter(&mut self) {
+        if self.whoami.is_none() {
+            self.set_whoami();
+            assert!(self.whoami.is_some());
+        }
+        if self.target_graph.is_none() {
+            self.create_target_graph();
+        }
+        assert!(self.whoami.is_some());
+    }
+
+    pub fn set_whoami(&mut self) {
         let my_node_wrapped = self.filter_state.get("node.metadata.WORKLOAD_NAME");
         if my_node_wrapped.is_none() {
             print!(
                 "WARNING: filter was initialized without envoy properties and thus cannot function"
             );
-            return to_return;
+            return;
         }
         let my_node = my_node_wrapped.unwrap().string_data.clone().unwrap();
+        self.whoami = Some(my_node);
+        assert!(self.whoami.is_some());
+    }
 
-        // 1. Do I need to put any udf variables/objects in?
-
-        // 2. Include any relevant node attributes
-        let mut data_to_append: String;
-        let mut data_key: String;
-        let mut me;
-
-        me = my_node.clone();
-        me.insert_str(0, " ");
-        me.push_str(".");
-
-        data_key = vec![
-            String::from("node"),
-            String::from("metadata"),
-            String::from("WORKLOAD_NAME"),
-        ]
-        .join(".");
-        data_to_append = self.filter_state[&data_key]
-            .string_data
-            .as_ref()
-            .unwrap()
-            .to_string();
-        me.push_str(&data_key);
-        me.push_str("==");
-        me.push_str(&data_to_append);
-        me.push_str(",");
-
-        if mod_rpc.headers.contains_key(&"properties".to_string()) {
-            mod_rpc
-                .headers
-                .get_mut(&"properties".to_string())
-                .unwrap()
-                .push_str(&me);
-        } else {
-            mod_rpc.headers.insert("properties".to_string(), me);
+    pub fn store_headers(&mut self, x: Rpc) {
+        // store path as well as properties
+        let prop_str = format!("{uid}_properties_path", uid = x.uid);
```

Comment:
Is this the code that is meant to go into the envoy filter, too?

Reply:
Yes.  It stores the path and any other collected properties using the form <uid>_properties_<property_name> in the envoy shared data.